### PR TITLE
Add `construct_batch` and `transform_program` functions to workflow module

### PR DESCRIPTION
**Pull request by albi3ro**
See #5058  for a prototype of the end state.

This PR adds two functions `qml.workflow.construct_batch` and `qml.workflow.transform_program`.

`transform_program(qnode, level)` takes a qnode and a "level" and returns a transform program.  The level keyword argument will be used through `construct_batch`, `draw`, `draw_mpl`, and `specs` and indicates a selection from the full transform program.

`construct_batch(qnode, level)(*args, **kwargs)`  takes a qnode and a level and returns a callable with the same signature as the qnode.  It then applies the transform program corresponding to `level`. 

Additional minor changes and helpers:

* `qml.transforms.core.expand_fn_transform`: In order to place `qml.Device.expand_fn` into the transform program easily, I added a quick function to convert from a tape->tape function to a transform.

* `TransformContainer.__repr__`: This just makes my life easier when working with transforms.

* Default the `is_informative` keyword argument to `False` instead of `None` when creating a transform.  This way a transform created with `TransformContainer` has the same default value of `is_informative`. This was causing me headaches when comparing transforms in testing.

* Slicing into a `TransformProgram` with a slice object returns another `TransformProgram` instead of a list. Ex `prog[0:4]`

* Defined `__contains__` for `TransformProgram`. This just made my life easier when playing around and testing things out.  example: `qml.compile in my_program`.

* Added some documentation on defined dunder methods in `TransformProgram`

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -85,6 +85,11 @@
 * Raise a more informative error when calling `adjoint_jacobian` with trainable state-prep operations.
   [(#5026)](https://github.com/PennyLaneAI/pennylane/pull/5026)
 
+* Adds `qml.workflow.transform_program` and `qml.workflow.construct_batch` to inspect the transform program and batch of tapes
+  at different stages.
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:89]:
```suggestion
  at different stages.
  [(#5084)](https://github.com/PennyLaneAI/pennylane/pull/5084)
```

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -85,6 +85,11 @@
...
+
+* Adds a `qml.transforms.core.expand_fn_transform` that converts a tape-> tape function into a transform.
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:91]:
```suggestion
* Adds a `qml.transforms.core.expand_fn_transform` that converts a tape-to-tape function into a transform.
  [(#5084)](https://github.com/PennyLaneAI/pennylane/pull/5084)
```

File: pennylane/transforms/core/transform.py [hunk 3]
```diff
@@ -50,7 +51,7 @@ def transform(
             ``quantum_transform``.
         classical_cotransform (Callable): A classical co-transform is a function to post-process the classical
             jacobian and the quantum jacobian and has the signature: ``my_cotransform(qjac, cjac, tape) -> tensor_like``
-        is_informative (bool): Whether or not a transform is informative. If true the transform is queued at the end
+        is_informative=False (bool): Whether or not a transform is informative. If true the transform is queued at the end
```
Comment (trbromley) [inline pennylane/transforms/core/transform.py:54]: Do we normally do this for defaults in the docstrings?

File: pennylane/transforms/core/transform.py [hunk 4]
```diff
@@ -177,15 +178,45 @@ def qnode_circuit(a):
...
+
+def null_postprocessing(results):
+    """A postprocessing function with null behavior."""
+    return results[0]
+
+
+def expand_fn_transform(
+    expand_fn: Callable[["pennylane.tape.QuantumTape"], "pennylane.tape.QuantumTape"]
+) -> "TransformDispatcher":
+    """Construct a transform from a tape-> tape function.
```
Comment (trbromley) [inline pennylane/transforms/core/transform.py:201]:
```suggestion
    """Construct a transform from a tape-to-tape function.
```

File: pennylane/transforms/core/transform.py [hunk 5]
```diff
@@ -177,15 +178,45 @@ def qnode_circuit(a):
...
+
+    Args:
+        expand_fn (Callable): a function from a single tape to a single tape
```
Comment (trbromley) [inline pennylane/transforms/core/transform.py:204]: Do we need to mention that the input single tape is within a length-1 list/tuple?

File: pennylane/transforms/core/transform.py [hunk 6]
```diff
@@ -177,15 +178,45 @@ def qnode_circuit(a):
...
+    Returns:
+
+        .TransformDispatcher: Returns a transform dispatcher object that that can transform any
+        circuit-like object in PennyLane.
+
+    >>> device = qml.device('default.qubit.legacy', wires=2)
+    >>> my_transform = qml.transforms.core.expand_fn_transform(device.expand_fn)
+    >>> my_transform
+    <transform: expand_fn>
+
```
Comment (trbromley) [inline pennylane/transforms/core/transform.py:215]:
```suggestion
```

File: pennylane/transforms/core/transform.py [hunk 7]
```diff
@@ -177,15 +178,45 @@ def qnode_circuit(a):
...
+    Returns:
+
+        .TransformDispatcher: Returns a transform dispatcher object that that can transform any
+        circuit-like object in PennyLane.
+
+    >>> device = qml.device('default.qubit.legacy', wires=2)
+    >>> my_transform = qml.transforms.core.expand_fn_transform(device.expand_fn)
+    >>> my_transform
+    <transform: expand_fn>
```
Comment (trbromley) [inline pennylane/transforms/core/transform.py:214]: To check I understand - the main utility of this new function is that you don't need to worry about the postprocessing function?

**Review by trbromley — COMMENTED** [review]
Thanks @albi3ro! This looks great. I think it's still WIP so I'll come back to look at this later.

File: pennylane/transforms/core/transform.py (hunk 5 above)
Comment (albi3ro) [reply pennylane/transforms/core/transform.py:204]: Single tape is a `QuantumScript`

File: doc/_static/transforms_order.png
Comment (lillian542) [inline doc/_static/transforms_order.png:1]: should gradient and device be capitalized like the other things?

Comment (albi3ro) [reply doc/_static/transforms_order.png:1]: or remove capitalization everywhere? Then we can match the `"gradient"` and `"device"` special points.

File: pennylane/workflow/construct_batch.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+        )
+    if isinstance(qnode.device, qml.devices.Device):
+        config = _make_execution_config(qnode)
+        return program + qnode.device.preprocess(config)[0]
+    program.add_transform(qml.transform(qnode.device.batch_transform))
+    program.add_transform(qml.transforms.core.expand_fn_transform(qnode.device.expand_fn))
+    return program
+
+
+def transform_program(qnode: "QNode", level=None) -> "qml.transforms.core.TransformProgram":
```
Comment (lillian542) [inline pennylane/workflow/construct_batch.py:39]: I wonder if this could be `get_transform_program` instead? To me `transform_program` sounds like a function that constructs a `TransformProgram`, rather than retrieving one from a `qnode`.

**Review by lillian542 — APPROVED** [review]
Looks great to me - `transform_program` and `construct_batch` both look super useful 🚀

File: pennylane/workflow/construct_batch.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+        """Create a batch of tapes and a post processing function."""
+        if "shots" in inspect.signature(qnode.func).parameters:
+            shots = _get_device_shots(qnode.device)
+        else:
+            shots = kwargs.pop("shots", _get_device_shots(qnode.device))
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qnode.func(*args, **kwargs)
+
+        initial_tape = qml.tape.QuantumScript.from_queue(q, shots)
```
Comment (timmysilv) [inline pennylane/workflow/construct_batch.py:265]: style nit-pick:
```suggestion
        initial_tape = qml.tape.make_qscript(qnode.func, shots=shots)(*args, **kwargs)
```

**Review by timmysilv — COMMENTED** [review]
really excited to have this in 🍾  most of my comments are about style or docs, lmk what you think

File: pennylane/workflow/construct_batch.py (hunk 8 above)
Comment (albi3ro) [reply pennylane/workflow/construct_batch.py:39]: That was something I was worrying about too.  I'll switch to `get_transform_program`.

File: pennylane/transforms/core/transform_program.py [hunk 10]
```diff
@@ -172,7 +210,10 @@ def push_back(self, transform_container: TransformContainer):
 
         # Program can only contain one informative transform and at the end of the program
         if self.has_final_transform:
-            raise TransformError("The transform program already has a terminal transform.")
+            if transform_container.final_transform:
+                raise TransformError("The transform program already has a terminal transform.")
+            self._transform_program.insert(-1, transform_container)
+            return
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/transforms/core/transform_program.py:216]: This is the same behaviour as adding a transform program with `+`.  If its problematic to add more transforms at the end of a transform program when a final transform exists, we also shouldn't be able to add in a transform program to the end.  


I'm fine with either decision, but we should be consistent.  We would then also have to figure out when gradient/ device/ internal transforms are run.

File: pennylane/transforms/core/transform.py (hunk 5 above)
Comment (trbromley) [reply pennylane/transforms/core/transform.py:204]: What I mean is that
> a function from a single tape to a single tape
doesn't feel like it matches the type hint:
```
Callable[["pennylane.tape.QuantumTape"], "pennylane.tape.QuantumTape"]
```

Comment (timmysilv) [reply pennylane/transforms/core/transform.py:204]: much to my disappointment, that is [the officially accepted Callable typing syntax](https://peps.python.org/pep-0677/)

File: pennylane/workflow/construct_batch.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+    """Extract a transform program at a designated level.
+
+    Args:
+        qnode (QNode): the qnode to get the transform program for.
+        level (None, str, int, slice): And indication
+
+            * ``None``: use the full transform program
+            * ``str``: Acceptable keys are ``"user"``, ``"device"``, and ``"gradient"``
+            * ``int``: How many transforms to include, starting from the front of the program
+            * ``slice``: a slice to select out components of the transform program.
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/workflow/construct_batch.py:49]: So yes, there are a lot of different ways we can designated the level, but I would call the main responsibility of this function is interpreting what the different levels mean.

So this will be the one source of truth for interpreting these designations.  `construct_batch` then relies on this function, and `draw`, `draw_mpl`, and `specs` will rely on `construct_batch`.

The most important level designations are the special strings and integers.  I added `None` and `slice` because they were an easy way to just tie into Python's existing slicing logic. I don't see `None` and `slice` being used as much in the wild.

We could instead just add a `level_to_slice` function, but I think this is a little more friendly.

File: pennylane/transforms/core/transform.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/transforms/core/transform.py:54]: We could potentially move the option keyword arguments to a "Keyword Args" section instead of using the "Args" section.

File: pennylane/transforms/core/transform.py [hunk 12]
```diff
@@ -177,15 +178,44 @@ def qnode_circuit(a):
...
     )
-    return dispatcher
+
+
+def null_postprocessing(results):
+    """A postprocessing function with null behavior."""
+    return results[0]
+
+
+def expand_fn_transform(
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/transforms/core/transform.py:198]: I'll add the `*args, **kwargs`.

Right now, I mostly see this as a patch for things that we can't easily move to the new system, rather than a recommended tool.  

I'm ok with either the name `single_tape_transform` or `expand_fn_transform`. Any opinions @trbromley ?

File: pennylane/devices/default_qubit.py [hunk 13]
```diff
@@ -159,6 +159,16 @@ def adjoint_state_measurements(
     )
 
 
+def adjoint_ops(op: qml.operation.Operator) -> bool:
+    """Specify whether or not an Operator is supported by adjoint differentiation."""
+    return op.num_params == 0 or op.num_params == 1 and op.has_generator
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/devices/default_qubit.py:164]: adding.

File: pennylane/transforms/core/transform.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/transforms/core/transform.py:198]: Good point about `*args, **kwargs`.

The purpose behind this was more for patching pre-existing code we can't easily switch.  I'm not sure we should particularly recommend it or use it places where we can use the alternative.

File: pennylane/workflow/construct_batch.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/workflow/construct_batch.py]: consider me convinced. plus it's all localized to a short, singular function so it's not that bad re: maintenance. I even thought "what about a function like `level_to_slice`?" but that's effectively what this is 😄

File: pennylane/transforms/core/transform.py (hunk 5 above)
Comment (albi3ro) [reply pennylane/transforms/core/transform.py:202]: eliminated the specifics of the callable signature when making it so `expand_fn` can take any additional args and kwargs.

File: pennylane/transforms/core/transform.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/transforms/core/transform.py]: Note this is something external users writing transforms will want/have wanted as well, so definitely something worth considering from a user POV.

For a researcher, the most ideal UI would simply be 

```python
@qml.transform
def my_transform(tape):
    new_ops = some_stuff(tape)
    new_tape = type(tape)(new_ops, tape.measurements, shots=tape.shots)
    return new_tape
```

That is, if no post-processing function/list of tapes is provided, it is treated as a single tape transform . It's a tradeoff --- more complex logic which differs depending on the return form, at the advantage of a single decorator the user needs to know about.

Out of scope for this PR, but something I wanted to mention for future consideration (cc @trbromley).

**Comment by timmysilv** [comment]
I'd like some finality on what to do with `expand_fn_transform` before this merges, but otherwise I think it's good to go. I'm fine to not make a public solution to the single-tape transform in this PR, but if that's the case, I would prefer that we keep this one private until we have a proposed UI for it

File: pennylane/transforms/core/transform.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/transforms/core/transform.py]: For now, what name do we want for this function and what import level do we make it accessible at?

[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/transforms/core/transform.py]: I'm removing this function from the transforms folder and making it private to `construct_batch.py`

**Review by timmysilv — APPROVED** [review]
love it

File: pennylane/transforms/core/transform.py (hunk 3 above)
Comment (trbromley) [reply pennylane/transforms/core/transform.py:53]: I'm still a bit unsure here on what is special about `is_informative` for us to add `=False`

File: pennylane/workflow/construct_batch.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+        qnode (QNode): the qnode to get the transform program for.
+        level (None, str, int, slice): And indication of what transforms to use from the full program.
+
+            * ``None``: use the full transform program
+            * ``str``: Acceptable keys are ``"user"``, ``"device"``, ``"top"`` and ``"gradient"``
+            * ``int``: How many transforms to include, starting from the front of the program
+            * ``slice``: a slice to select out components of the transform program.
+
+    Returns:
+        TransformProrgram: the transform program corresponding to the requested level.
```
Comment (trbromley) [inline pennylane/workflow/construct_batch.py:82]:
```suggestion
        TransformProgram: the transform program corresponding to the requested level.
```

File: pennylane/workflow/construct_batch.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+        dev = qml.device('default.qubit')
+
+        @qml.metric_tensor # final transform
+        @qml.transforms.merge_rotations # transform 2
+        @qml.transforms.cancel_inverses # transform 1
+        @qml.qnode(dev, diff_method="parameter-shift", shifts=np.pi / 4)
+        def circuit():
+            return qml.expval(qml.PauliZ(0))
+
+    By default, we get the full transform program. This can be manaully specified by ``level=None``.
```
Comment (trbromley) [inline pennylane/workflow/construct_batch.py:107]:
```suggestion
    By default, we get the full transform program. This can be manually specified by ``level=None``.
```

File: pennylane/workflow/construct_batch.py [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+            * ``None``: use the full transform program
+            * ``str``: Acceptable keys are ``"user"``, ``"device"``, ``"top"`` and ``"gradient"``
+            * ``int``: How many transforms to include, starting from the front of the program
+            * ``slice``: a slice to select out components of the transform program.
+
+    Returns:
+        TransformProrgram: the transform program corresponding to the requested level.
+
+
+    The transforms are organized as:
```
Comment (trbromley) [inline pennylane/workflow/construct_batch.py:85]: Should we embed the following in usage details?

File: pennylane/workflow/construct_batch.py [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+    Or what the parameter shift gradient transform will be applied to:
+
+    >>> batch, fn = construct_batch(circuit, level="gradient")(1.23)
+    >>> batch[0].circuit
+    [RY(tensor(1., requires_grad=True), wires=[1]),
+    RX(tensor(2., requires_grad=True), wires=[0]),
+    expval(  (1) [X0]
+    + (1) [Y0])]
+
+    We can inpsect what was directly captured from the qfunc with ``level=0``.
```
Comment (trbromley) [inline pennylane/workflow/construct_batch.py:245]:
```suggestion
    We can inspect what was directly captured from the qfunc with ``level=0``.
```

**Review by trbromley — APPROVED** [review]
Nice! Approved, subject to my comments.

To check I understand, this will be used for the drawing and specs functionality?

File: doc/releases/changelog-dev.md [hunk 18]
```diff
@@ -139,6 +139,13 @@
 * Raise a more informative error when calling `adjoint_jacobian` with trainable state-prep operations.
   [(#5026)](https://github.com/PennyLaneAI/pennylane/pull/5026)
 
+* Adds `qml.workflow.get_transform_program` and `qml.workflow.construct_batch` to inspect the transform program and batch of tapes
+  at different stages.
+  [(#5084)](https://github.com/PennyLaneAI/pennylane/pull/5084)
+
+* Adds a `qml.transforms.core.expand_fn_transform` that converts a tape-to-tape function into a transform.
+  [(#5084)](https://github.com/PennyLaneAI/pennylane/pull/5084)
+
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:148]:
```suggestion
```

File: pennylane/transforms/core/transform.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/transforms/core/transform.py:53]: I'm separating it out into a "Arrgs" and "Keyword Args" section to include the default values for the Keyword Args.

**Comment by albi3ro** [comment]
> Nice! Approved, subject to my comments.
> 
> To check I understand, this will be used for the drawing and specs functionality?

Yes this will be for drawing and specs.

File: pennylane/transforms/core/transform.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/transforms/core/transform.py:54]: Basically, I didn't realize that the default value originally was `None` and it caused me a headache. Hence why I want to be a bit more explicit about what the default values are.
